File: java/src/org/smc/inputmethod/indic/settings/LearnedWordsFragment.java
/*
 * Copyright 2026, Jishnu Mohan <jishnu7@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.smc.inputmethod.indic.settings;

import android.content.Context;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodSubtype;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.android.inputmethod.latin.R;
import com.android.inputmethod.latin.RichInputMethodManager;
import com.android.inputmethod.latin.common.LocaleUtils;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import org.smc.inputmethod.indic.personalization.PersonalizationHelper;
import org.smc.inputmethod.indic.personalization.UserHistoryDictionary;

import java.io.File;
import java.text.Collator;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Locale;

/**
 * Lists the words the keyboard learned from typing (the user history dictionaries), one
 * language at a time via filter chips, with search, per-word delete and a delete-all action.
 */
public final class LearnedWordsFragment extends Fragment {
    private static final String DICT_FILE_PREFIX = "UserHistoryDictionary.";
    private static final String DICT_FILE_SUFFIX = ".dict";
    private static final int MENU_DELETE_ALL = Menu.FIRST;

    private ChipGroup mLanguages;
    private EditText mSearch;
    private TextView mEmpty;
    private RecyclerView mList;
    private WordsAdapter mAdapter;

    private Locale mSelectedLocale;
    private final ArrayList<String> mWords = new ArrayList<>();

    @Override
    public View onCreateView(final LayoutInflater inflater, final ViewGroup container,
            final Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        final View root = inflater.inflate(R.layout.fragment_learned_words, container, false);
        mLanguages = root.findViewById(R.id.learned_words_languages);
        mSearch = root.findViewById(R.id.learned_words_search);
        mEmpty = root.findViewById(R.id.learned_words_empty);
        mList = root.findViewById(R.id.learned_words_list);
        mList.setLayoutManager(new LinearLayoutManager(requireContext()));
        mAdapter = new WordsAdapter();
        mList.setAdapter(mAdapter);
        mSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count,
                    final int after) {}

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before,
                    final int count) {}

            @Override
            public void afterTextChanged(final Editable s) {
                mAdapter.filter(s.toString());
            }
        });
        buildLanguageChips();
        return root;
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getActivity() instanceof AppCompatActivity) {
            final ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
            if (actionBar != null) {
                actionBar.setTitle(R.string.privacy_learned_typing);
            }
        }
    }

    @Override
    public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        menu.add(Menu.NONE, MENU_DELETE_ALL, Menu.NONE, R.string.learned_words_delete_all)
                .setIcon(R.drawable.ic_settings_delete)
                .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        if (item.getItemId() != MENU_DELETE_ALL) {
            return super.onOptionsItemSelected(item);
        }
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle(R.string.privacy_learned_typing)
                .setMessage(R.string.privacy_learned_typing_confirm)
                .setPositiveButton(R.string.privacy_delete, (dialog, which) -> {
                    PersonalizationHelper.removeAllUserHistoryDictionaries(requireContext());
                    buildLanguageChips();
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
        return true;
    }

    // The user history dictionary lives in memory between flushes, so the enabled keyboard
    // languages are the source of truth; on-disk dictionaries cover languages since disabled.
    private ArrayList<Locale> findLocalesWithHistory() {
        final LinkedHashMap<String, Locale> locales = new LinkedHashMap<>();
        RichInputMethodManager.init(requireContext());
        for (final InputMethodSubtype subtype : RichInputMethodManager.getInstance()
                .getMyEnabledInputMethodSubtypeList(true /* allowsImplicitlySelectedSubtypes */)) {
            final Locale locale = LocaleUtils.constructLocaleFromString(subtype.getLocale());
            locales.put(locale.toString(), locale);
        }
        final File[] dictFiles = requireContext().getFilesDir().listFiles((dir, name) ->
                name.startsWith(DICT_FILE_PREFIX) && name.endsWith(DICT_FILE_SUFFIX));
        if (dictFiles != null) {
            for (final File file : dictFiles) {
                final String localeString = file.getName().substring(DICT_FILE_PREFIX.length(),
                        file.getName().length() - DICT_FILE_SUFFIX.length());
                locales.put(localeString, LocaleUtils.constructLocaleFromString(localeString));
            }
        }
        final ArrayList<Locale> sorted = new ArrayList<>(locales.values());
        Collections.sort(sorted, (a, b) -> a.toString().compareTo(b.toString()));
        return sorted;
    }

    private void buildLanguageChips() {
        final Context context = requireContext();
        final ArrayList<Locale> locales = findLocalesWithHistory();
        mLanguages.removeAllViews();
        mSelectedLocale = null;
        if (locales.isEmpty()) {
            showWords(new ArrayList<>());
            return;
        }
        for (final Locale locale : locales) {
            final Chip chip = new Chip(context);
            chip.setText(locale.getDisplayName(locale));
            chip.setCheckable(true);
            chip.setTag(locale);
            mLanguages.addView(chip);
        }
        mLanguages.setOnCheckedStateChangeListener((group, checkedIds) -> {
            if (checkedIds.isEmpty()) {
                return;
            }
            final Chip chip = group.findViewById(checkedIds.get(0));
            loadWords((Locale) chip.getTag());
        });
        ((Chip) mLanguages.getChildAt(0)).setChecked(true);
    }

    private void loadWords(final Locale locale) {
        mSelectedLocale = locale;
        final UserHistoryDictionary dictionary = PersonalizationHelper.getUserHistoryDictionary(
                requireContext(), locale, null /* account */);
        dictionary.getAllWordsAsync(words -> {
            if (!isAdded() || !locale.equals(mSelectedLocale)) {
                return;
            }
            final Collator collator = Collator.getInstance(locale);
            Collections.sort(words, collator);
            showWords(words);
        });
    }

    private void showWords(final ArrayList<String> words) {
        mWords.clear();
        mWords.addAll(words);
        mAdapter.filter(mSearch.getText().toString());
    }

    private void updateEmptyState() {
        final boolean empty = mAdapter.getItemCount() == 0;
        mEmpty.setVisibility(empty ? View.VISIBLE : View.GONE);
        mList.setVisibility(empty ? View.GONE : View.VISIBLE);
    }

    private void deleteWord(final String word) {
        if (mSelectedLocale == null) {
            return;
        }
        PersonalizationHelper.getUserHistoryDictionary(requireContext(), mSelectedLocale,
                null /* account */).removeUnigramEntryDynamically(word);
        mWords.remove(word);
        mAdapter.filter(mSearch.getText().toString());
    }

    private final class WordsAdapter extends RecyclerView.Adapter<WordViewHolder> {
        private final ArrayList<String> mFiltered = new ArrayList<>();

        void filter(final String query) {
            mFiltered.clear();
            if (TextUtils.isEmpty(query)) {
                mFiltered.addAll(mWords);
            } else {
                final Locale locale = (mSelectedLocale != null) ? mSelectedLocale
                        : Locale.getDefault();
                final String needle = query.toLowerCase(locale);
                for (final String word : mWords) {
                    if (word.toLowerCase(locale).contains(needle)) {
                        mFiltered.add(word);
                    }
                }
            }
            notifyDataSetChanged();
            updateEmptyState();
        }

        @Override
        public WordViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
            final View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.learned_word_item, parent, false);
            return new WordViewHolder(view);
        }

        @Override
        public void onBindViewHolder(final WordViewHolder holder, final int position) {
            final String word = mFiltered.get(position);
            holder.mWord.setText(word);
            final boolean top = position == 0;
            final boolean bottom = position == getItemCount() - 1;
            holder.itemView.setBackgroundResource(cardBackground(top, bottom));
            holder.mDelete.setOnClickListener(v -> deleteWord(word));
        }

        @Override
        public int getItemCount() {
            return mFiltered.size();
        }
    }

    private static int cardBackground(final boolean top, final boolean bottom) {
        if (top && bottom) return R.drawable.pref_card_single;
        if (top) return R.drawable.pref_card_top;
        if (bottom) return R.drawable.pref_card_bottom;
        return R.drawable.pref_card_middle;
    }

    private static final class WordViewHolder extends RecyclerView.ViewHolder {
        final TextView mWord;
        final View mDelete;

        WordViewHolder(final View itemView) {
            super(itemView);
            mWord = itemView.findViewById(R.id.learned_word);
            mDelete = itemView.findViewById(R.id.learned_word_delete);
        }
    }
}

File: java/src/org/smc/inputmethod/indic/settings/LearnedWordsFragment.kt
/*
 * Copyright 2026, Jishnu Mohan <jishnu7@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.smc.inputmethod.indic.settings

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.TextView

import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.android.inputmethod.latin.R
import com.android.inputmethod.latin.RichInputMethodManager
import com.android.inputmethod.latin.common.LocaleUtils
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.google.android.material.dialog.MaterialAlertDialogBuilder

import org.smc.inputmethod.indic.personalization.PersonalizationHelper

import java.text.Collator
import java.util.Locale

/**
 * Lists the words the keyboard learned from typing (the user history dictionaries), one language
 * at a time via filter chips, with search, per-word delete and a delete-all action.
 */
@Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
class LearnedWordsFragment : Fragment() {

    private lateinit var languages: ChipGroup
    private lateinit var search: EditText
    private lateinit var empty: TextView
    private lateinit var list: RecyclerView
    private lateinit var adapter: WordsAdapter

    private var selectedLocale: Locale? = null
    private val words = ArrayList<String>()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        setHasOptionsMenu(true)
        val root = inflater.inflate(R.layout.fragment_learned_words, container, false)
        languages = root.findViewById(R.id.learned_words_languages)
        search = root.findViewById(R.id.learned_words_search)
        empty = root.findViewById(R.id.learned_words_empty)
        list = root.findViewById(R.id.learned_words_list)
        list.layoutManager = LinearLayoutManager(requireContext())
        adapter = WordsAdapter()
        list.adapter = adapter
        search.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable) {
                adapter.filter(s.toString())
            }
        })
        buildLanguageChips()
        return root
    }

    override fun onResume() {
        super.onResume()
        (activity as? AppCompatActivity)?.supportActionBar?.setTitle(R.string.privacy_learned_typing)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        menu.add(Menu.NONE, MENU_DELETE_ALL, Menu.NONE, R.string.learned_words_delete_all)
            .setIcon(R.drawable.ic_settings_delete)
            .setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId != MENU_DELETE_ALL) {
            return super.onOptionsItemSelected(item)
        }
        MaterialAlertDialogBuilder(requireContext())
            .setTitle(R.string.privacy_learned_typing)
            .setMessage(R.string.privacy_learned_typing_confirm)
            .setPositiveButton(R.string.privacy_delete) { _, _ ->
                PersonalizationHelper.removeAllUserHistoryDictionaries(requireContext())
                buildLanguageChips()
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
        return true
    }

    // The user history dictionary lives in memory between flushes, so the enabled keyboard
    // languages are the source of truth; on-disk dictionaries cover languages since disabled.
    private fun findLocalesWithHistory(): List<Locale> {
        val locales = LinkedHashMap<String, Locale>()
        RichInputMethodManager.init(requireContext())
        for (subtype in RichInputMethodManager.getInstance()
            .getMyEnabledInputMethodSubtypeList(true /* allowsImplicitlySelectedSubtypes */)) {
            val locale = LocaleUtils.constructLocaleFromString(subtype.locale)
            locales[locale.toString()] = locale
        }
        val dictFiles = requireContext().filesDir.listFiles { _, name ->
            name.startsWith(DICT_FILE_PREFIX) && name.endsWith(DICT_FILE_SUFFIX)
        }
        dictFiles?.forEach { file ->
            val localeString =
                file.name.substring(DICT_FILE_PREFIX.length, file.name.length - DICT_FILE_SUFFIX.length)
            locales[localeString] = LocaleUtils.constructLocaleFromString(localeString)
        }
        return locales.values.sortedBy { it.toString() }
    }

    private fun buildLanguageChips() {
        val context = requireContext()
        val locales = findLocalesWithHistory()
        languages.removeAllViews()
        selectedLocale = null
        if (locales.isEmpty()) {
            showWords(ArrayList())
            return
        }
        for (locale in locales) {
            val chip = Chip(context)
            chip.text = locale.getDisplayName(locale)
            chip.isCheckable = true
            chip.tag = locale
            languages.addView(chip)
        }
        languages.setOnCheckedStateChangeListener { group, checkedIds ->
            if (checkedIds.isEmpty()) return@setOnCheckedStateChangeListener
            val chip = group.findViewById<Chip>(checkedIds[0])
            loadWords(chip.tag as Locale)
        }
        (languages.getChildAt(0) as Chip).isChecked = true
    }

    private fun loadWords(locale: Locale) {
        selectedLocale = locale
        val dictionary = PersonalizationHelper.getUserHistoryDictionary(
            requireContext(), locale, null /* account */
        )
        dictionary.getAllWordsAsync { fetched ->
            if (!isAdded || locale != selectedLocale) return@getAllWordsAsync
            val collator = Collator.getInstance(locale)
            fetched.sortWith { a, b -> collator.compare(a, b) }
            showWords(fetched)
        }
    }

    private fun showWords(fetched: ArrayList<String>) {
        words.clear()
        words.addAll(fetched)
        adapter.filter(search.text.toString())
    }

    private fun updateEmptyState() {
        val isEmpty = adapter.itemCount == 0
        empty.visibility = if (isEmpty) View.VISIBLE else View.GONE
        list.visibility = if (isEmpty) View.GONE else View.VISIBLE
    }

    private fun deleteWord(word: String) {
        val locale = selectedLocale ?: return
        PersonalizationHelper.getUserHistoryDictionary(requireContext(), locale, null /* account */)
            .removeUnigramEntryDynamically(word)
        words.remove(word)
        adapter.filter(search.text.toString())
    }

    private inner class WordsAdapter : RecyclerView.Adapter<WordViewHolder>() {
        private val filtered = ArrayList<String>()

        fun filter(query: String) {
            filtered.clear()
            if (query.isEmpty()) {
                filtered.addAll(words)
            } else {
                val locale = selectedLocale ?: Locale.getDefault()
                val needle = query.lowercase(locale)
                words.filterTo(filtered) { it.lowercase(locale).contains(needle) }
            }
            notifyDataSetChanged()
            updateEmptyState()
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): WordViewHolder {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.learned_word_item, parent, false)
            return WordViewHolder(view)
        }

        override fun onBindViewHolder(holder: WordViewHolder, position: Int) {
            val word = filtered[position]
            holder.wordText.text = word
            val top = position == 0
            val bottom = position == itemCount - 1
            holder.itemView.setBackgroundResource(cardBackground(top, bottom))
            holder.deleteButton.setOnClickListener { deleteWord(word) }
        }

        override fun getItemCount(): Int = filtered.size
    }

    private class WordViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val wordText: TextView = itemView.findViewById(R.id.learned_word)
        val deleteButton: View = itemView.findViewById(R.id.learned_word_delete)
    }

    companion object {
        private const val DICT_FILE_PREFIX = "UserHistoryDictionary."
        private const val DICT_FILE_SUFFIX = ".dict"
        private val MENU_DELETE_ALL = Menu.FIRST

        private fun cardBackground(top: Boolean, bottom: Boolean): Int = when {
            top && bottom -> R.drawable.pref_card_single
            top -> R.drawable.pref_card_top
            bottom -> R.drawable.pref_card_bottom
            else -> R.drawable.pref_card_middle
        }
    }
}
